elements/lisk-transactions/src/13_vote_transaction.ts

```diff
@@ -0,0 +1,450 @@
+						'.asset.votes.amount',
+					),
+				);
+			}
+			if (vote.amount % TEN_UNIT !== BigInt(0)) {
+				errors.push(
+					new TransactionError(
+						'Amount should be multiple of 10 * 10^8',
+						this.id,
+						'.asset.votes.amount',
+						vote.amount.toString(),
+					),
+				);
+			}
+		}
+		if (upvoteCount > MAX_VOTE) {
+			errors.push(
+				new TransactionError(
+					'Upvote can only be casted upto 10',
+					this.id,
+					'.asset.votes',
+					upvoteCount,
+					'10',
+				),
+			);
+		}
+		if (downvoteCount > MAX_VOTE) {
+			errors.push(
+				new TransactionError(
+					'Downvote can only be casted upto 10',
+					this.id,
+					'.asset.votes',
+					upvoteCount,
+					'10',
+				),
+			);
+		}
+		if (addressSet.size !== this.asset.votes.length) {
+			errors.push(
+				new TransactionError(
+					'Delegate address must be unique',
+					this.id,
+					'.asset.votes.delegateAddress',
+				),
+			);
+		}
+
+		return errors;
+	}
+
+	protected async applyAsset(
+		store: StateStore,
+	): Promise<ReadonlyArray<TransactionError>> {
+		// Only order should be change, so no need to copy object itself
+		const assetCopy = [...this.asset.votes];
+		// Sort by acending amount
+		assetCopy.sort((a, b) => {
+			const diff = a.amount - b.amount;
+			if (diff > BigInt(0)) {
+				return 1;
+			}
+			if (diff < BigInt(0)) {
+				return -1;
+			}
+
+			return 0;
+		});
+		const errors = [];
+
+		for (const vote of assetCopy) {
+			const sender = await store.account.get(this.senderId);
+			const delegate = await store.account.get(vote.delegateAddress);
+
+			if (!delegate.username) {
+				errors.push(
+					new TransactionError(
+						'Voted delegate is not registered',
+						this.id,
+						'.asset.votes.delegateAddress',
+					),
+				);
+				continue;
+			}
+			if (vote.amount < BigInt(0)) {
+				const originalUpvoteIndex = sender.votes.findIndex(
+					senderVote => senderVote.delegateAddress === vote.delegateAddress,
+				);
+				if (originalUpvoteIndex < 0) {
+					errors.push(
+						new TransactionError(
+							'Cannot cast downvote to delegate who is not upvoted',
+							this.id,
+							'.asset.votes.delegateAddress',
+						),
+					);
+					continue;
+				}
+				sender.votes[originalUpvoteIndex].amount += vote.amount;
+				if (sender.votes[originalUpvoteIndex].amount < BigInt(0)) {
+					errors.push(
+						new TransactionError(
+							'Cannot downvote more than upvoted',
+							this.id,
+							'.asset.votes.amount',
+						),
+					);
+				}
+				// Delete entry when amount becomes 0
+				if (sender.votes[originalUpvoteIndex].amount === BigInt(0)) {
+					sender.votes = sender.votes.filter(
+						senderVote => senderVote.delegateAddress !== vote.delegateAddress,
+					);
+				}
+				// Create unlocking object
+				sender.unlocking.push({
+					delegateAddress: vote.delegateAddress,
+					amount: BigInt(-1) * vote.amount,
+					unvoteHeight: store.chain.lastBlockHeader.height + 1,
+				});
+				// Sort account.unlocking
```

Comment:
we should also account for the case when the delegate address is the same, the sorting should be lexicographical by (address, unvoteHeight, amount).